velox/exec/tests/MultiFragmentTest.cpp

```diff
@@ -1400,6 +1404,56 @@ TEST_F(MultiFragmentTest, taskTerminateWithPendingOutputBuffers) {
   task.reset();
 }
 
+TEST_F(MultiFragmentTest, taskTerminateWithProblematicRemainingRemoteSplits) {
+  // Start the task with 2 drivers.
+  auto probeData =
+      makeRowVector({"p_c0"}, {makeFlatVector<int64_t>({1, 2, 3})});
+  auto planNodeIdGenerator = std::make_shared<core::PlanNodeIdGenerator>();
+  core::PlanNodeId exchangeNodeId;
+  auto plan = PlanBuilder(planNodeIdGenerator)
+                  .values({probeData}, true)
+                  .hashJoin(
+                      {"p_c0"},
+                      {"c0"},
+                      PlanBuilder(planNodeIdGenerator)
+                          .exchange(rowType_)
+                          .capturePlanNodeId(exchangeNodeId)
+                          .planNode(),
+                      "",
+                      {"c0"})
+                  .partitionedOutput({}, 1)
+                  .planNode();
+  auto taskId = makeTaskId("final", 0);
+  auto task = makeTask(taskId, plan, 0);
+  Task::start(task, 2);
+
+  // Wait for all drivers to be blocked, so that the promises will be made.
+  bool allDriversBlocked = false;
+  while (!allDriversBlocked) {
+    allDriversBlocked = true;
+    task->testingVisitDrivers([&](Driver* driver) {
+      if (driver->isOnThread() ||
+          (driver->blockingReason() != BlockingReason::kWaitForSplit &&
+           driver->blockingReason() != BlockingReason::kWaitForProducer &&
+           driver->blockingReason() != BlockingReason::kWaitForJoinBuild)) {
+        allDriversBlocked = false;
+      }
+    });
+  }
+
+  // Add one problematic task and trigger Task::terminate.
+  task->addSplit(exchangeNodeId, remoteSplit(makeErrorTaskId("leaf", 0)));
+
```

Comment:
Note: the test has timing dependency here. We can't guarantee the second bad split can always be added to task without using testvalue. But it is okay and without this fix, the test will be flaky then.